### PR TITLE
rewrite file only when file data changed

### LINKED ISSUE #597: only print the file names of changed files
for now, when I run `js-beautify` from cli, it will print all the file names passed to `js-beautify`, but some are not changed.

So, it's better to print the names of `changed files only`.

### PR DESCRIPTION
**Pull request by coderhaoxin**
also fix #597

### CODE REVIEW

**Comment by bitwiseman** [comment]
Good start, thanks!  Now make it pass tests and also implement it python.

**Comment by coderhaoxin** [comment]
ok. soon

**Comment by coderhaoxin** [comment]
> also implement it python.

done

test seems failed on `master` too.
```
js-beautify output for /tmp/js-beautify-mkdir/example1-default.js was expected to be different
```

run the cli directly is ok, file created in `/tmp/js-beautify-mkdir`

**Comment by bitwiseman** [comment]
No, the problem is the test expects even unchanged files to be written out...

File: js/lib/cli.js [hunk 1]
```diff
@@ -293,24 +293,26 @@ function makePretty(code, config, outfile, callback) {
...
         process.exit(1);
     }
 
     if (outfile) {
-        try {
-            fs.writeFileSync(outfile, pretty, 'utf8');
-            logToStdout('beautified ' + path.relative(process.cwd(), outfile), config);
-        } catch (ex) {
-            onOutputError(ex);
+        if (code !== pretty) {
```
Comment (bitwiseman) [inline js/lib/cli.js]: Here's the problem.  You can't just not write the code if it is the same.  You need to check (around line 267) if the input and output path are the same.  If so, you can check here if the input and output code is unchanged and not change that file.  If the input path is empty, output path is empty (for piping cases), or if the input path differs from the output path you still have to write the output, even if it is unchanged.

File: js/lib/cli.js [hunk 2]
```diff
@@ -278,39 +278,41 @@ function processInputSync(filepath) {
...
 }
 
-function makePretty(code, config, outfile, callback) {
+function makePretty(code, config, filepath, outfile, callback) {
     try {
         var fileType = getOutputType(outfile, config.type);
         var pretty = beautify[fileType](code, config);
 
-        callback(null, pretty, outfile, config);
+        callback(null, pretty, filepath, outfile, config, code);
```
Comment (bitwiseman) [inline js/lib/cli.js]:
```js
if (!(data === pretty && filepath === outfile)) {
    callback(null, pretty, outfile, config);
} 
```

Then you can revert the changes in `writePretty`.

**Comment by bitwiseman** [comment]
Cool. We don't have a test to show that unchanged files are not written, but I can do that.  And all the tests pass.  Thanks!

**Comment by coderhaoxin** [comment]
:smile:

**Comment by blino** [comment]
The new tests seem to be OSX-specific, touch -A does not work with touch from GNU coreutils (Linux).

**Comment by bitwiseman** [comment]
@blino - please open a new issue, and if you are willing to contribute a fix that would be even better!
